src/devices/bus/epson_qx/cqgmem.cpp

```diff
@@ -0,0 +1,142 @@
+// license:BSD-3-Clause
+// copyright-holders:Brian Johnson
+/*******************************************************************
+ *
+ * commodity quote graphics 1 megabyte memory expansion
+ *
+ *******************************************************************/
+
+#include "emu.h"
+#include "cqgmem.h"
+
+//**************************************************************************
+//  CQGMEM DEVICE
+//**************************************************************************
+
+DEFINE_DEVICE_TYPE(EPSON_QX_OPTION_CQGMEM, bus::epson_qx::cqgmem_device, "option_cqgmem", "Epson QX-10 1MB Memory Expansion")
+
+namespace bus::epson_qx {
+
+static INPUT_PORTS_START( cqgmem )
+	PORT_START("IOBASE")
+	PORT_CONFNAME(0xf0, 0xf0, "IO Base Address Selection")
+	PORT_CONFSETTING(0x80, "&80")
+	PORT_CONFSETTING(0x90, "&90")
+	PORT_CONFSETTING(0xa0, "&A0")
+	PORT_CONFSETTING(0xb0, "&B0")
+	PORT_CONFSETTING(0xc0, "&C0")
+	PORT_CONFSETTING(0xd0, "&D0")
+	PORT_CONFSETTING(0xe0, "&E0")
+	PORT_CONFSETTING(0xf0, "&F0")
+INPUT_PORTS_END
+
+//-------------------------------------------------
+//  cqgmem_device - constructor
+//-------------------------------------------------
+cqgmem_device::cqgmem_device(const machine_config &mconfig, const char *tag, device_t *owner, uint32_t clock) :
+	device_t(mconfig, EPSON_QX_OPTION_CQGMEM, tag, owner, clock),
+	device_option_expansion_interface(mconfig, *this),
+	device_memory_interface(mconfig, *this),
+	m_ram(*this, RAM_TAG),
+	m_banks(*this, "bank%u", 0U),
+	m_iobase(*this, "IOBASE"),
+	m_space_config("xmem", ENDIANNESS_LITTLE, 8, 20, 0, address_map_constructor(FUNC(cqgmem_device::xmem_map), this)),
+	m_installed(false)
+{
+}
+
+//-------------------------------------------------
+//  device_input_ports - device-specific ports
+//-------------------------------------------------
+ioport_constructor cqgmem_device::device_input_ports() const
+{
+	return INPUT_PORTS_NAME( cqgmem );
+}
+
+//-------------------------------------------------
+//  device_add_mconfig - device-specific config
+//-------------------------------------------------
+void cqgmem_device::device_add_mconfig(machine_config &config)
+{
+	RAM(config, RAM_TAG).set_default_size("1M");
+}
```

Comment:
You shouldn’t be trying to create a device outside your device’s place in the device hierarchy.  Just use a `memory_share_creator` – this isn’t really appropriate use of `ram_device` anyway.